src/main/java/com/ververica/flink/table/gateway/deployment/ClusterDescriptorAdapter.java

```diff
@@ -0,0 +1,138 @@
+				try {
+					return clusterClient.getJobStatus(jobId).get(DEFAULT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+				} catch (InterruptedException | ExecutionException | TimeoutException e) {
+					LOG.error(String.format("Session: %s. Failed to fetch job status for job %s", sessionId, jobId), e);
+					throw new SqlGatewayException("Failed to fetch job status for job " + jobId, e);
+				}
+			});
+    }
+
+    /**
+     * Cancel the flink job.
+     */
+    public void cancelJob() {
+        LOG.info("Session: {}. Start to cancel job {}.", sessionId, jobId);
+        bridgeClientRequest(this.executionContext, jobId, sessionId, clusterClient -> {
+            try {
+                clusterClient.cancel(jobId).get();
+            } catch (Throwable t) {
+                // the job might has finished earlier
+            }
+            return null;
+        });
+    }
+
+    /**
+     * The reason of using ClusterClient instead of JobClient to retrieve a cluster is
+     * the JobClient can't know whether the job is finished on yarn-per-job mode.
+     *
+     * <p>If a job is finished, JobClient always get java.util.concurrent.TimeoutException
+     * when getting job status and canceling a job after job is finished.
+     * This method will throw org.apache.hadoop.yarn.exceptions.ApplicationNotFoundException
+     * when creating a ClusterClient if the job is finished. This is more user-friendly.
+     */
+    protected <R> R bridgeClientRequest(
+            ExecutionContext<ClusterID> executionContext,
+            JobID jobId,
+            String sessionId,
+            Function<ClusterClient<?>, R> function) {
+        // stop Flink job
+        try (final ClusterDescriptor<ClusterID> clusterDescriptor = executionContext.createClusterDescriptor()) {
+            try (ClusterClient<ClusterID> clusterClient =
+                         clusterDescriptor.retrieve(this.clusterID).getClusterClient()) {
+                // retrieve existing cluster
+                return function.apply(clusterClient);
+            } catch (Exception e) {
+                LOG.error(
+                        String.format("Session: %s, job: %s. Could not retrieve or create a cluster.", sessionId, jobId),
+                        e);
+                throw new SqlExecutionException("Could not retrieve or create a cluster.", e);
+            }
+        } catch (SqlExecutionException e) {
+            throw e;
+        } catch (Exception e) {
+            LOG.error(
+                    String.format("Session: %s, job: %s. Could not locate a cluster.", sessionId, jobId), e);
+            throw new SqlExecutionException("Could not locate a cluster.", e);
+        }
+    }
+
+    /**
+     * Set the flink job id.
+     */
+    public void setJobId(JobID jobId) {
+        this.jobId = jobId;
+    }
+
+    /**
+     * Check whether the current flink job status is GloballyTerminalState.
+     */
+    public abstract boolean isGloballyTerminalState();
```

Comment:
explain more for `GloballyTerminalState` (can refer to `JobStatus#isGloballyTerminalState`)